### PR TITLE
SQ-237/Convert onboarding and routing to Kotlin

### PR DESCRIPTION
**Pull request by rock3r**
This PR converts the onboarding and routing/navigation domains to Kotlin, too.

### CODE REVIEW

File: app/src/main/java/net/squanchy/navigation/FirstStartPersister.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    fun storeHasBeenStarted() {
+        preferences.edit()
+                .putBoolean(KEY_HAS_BEEN_STARTED_ALREADY, true)
+                .apply()
+    }
+
+    companion object {
+
+        private val KEY_HAS_BEEN_STARTED_ALREADY = "FirstStart.has_been_started_already"
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/navigation/FirstStartPersister.kt]: const

File: app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkNavigator.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import net.squanchy.navigation.Navigator
+import net.squanchy.support.lang.Optional
+
+internal class DeepLinkNavigator(private val navigator: Navigator) {
+
+    fun toSchedule(path: String) {
+        val pathSegments = extractSegments(path)
+
+        var dayId = Optional.absent<String>()
+        if (!pathSegments.isEmpty()) {
```
Comment (quiro91) [inline app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkNavigator.kt]: `pathSegments.isNotEmpty()`

**Review by quiro91 — APPROVED** [review]

File: app/src/main/java/net/squanchy/navigation/NavigationModule.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package net.squanchy.navigation
+
+import android.app.Activity
+import dagger.Module
+import dagger.Provides
+import net.squanchy.injection.ActivityContextModule
+
+@Module(includes = arrayOf(ActivityContextModule::class))
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/NavigationModule.kt:8]: isn't `includes` seen as a vararg in kotlin?

File: app/src/main/java/net/squanchy/navigation/Navigator.kt [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+internal class Navigator(private val activity: Activity, private val debugActivityIntentFactory: DebugActivityIntentFactory) {
+
+    fun toEventDetails(eventId: String) {
+        start(EventDetailsActivity.createIntent(activity, eventId))
+    }
+
+    fun toSpeakerDetails(speakerId: String) {
+        start(
+                SpeakerDetailsActivity.createIntent(activity, speakerId),
+                FLAG_ACTIVITY_SINGLE_TOP or FLAG_ACTIVITY_CLEAR_TOP
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/Navigator.kt:35]: I see why `or` but technically you're passing both flags, this reads weird :P

how about

```
infix fun Int.also(other: Int) = or(other)
```

?

File: app/src/main/java/net/squanchy/navigation/RoutingActivity.kt [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+class RoutingActivity : TypefaceStyleableActivity() {
+
+    private lateinit var deepLinkRouter: DeepLinkRouter
+    private lateinit var navigator: Navigator
+    private lateinit var onboarding: Onboarding
+    private lateinit var signInService: SignInService
+    private lateinit var firstStartPersister: FirstStartPersister
+
+    private val subscriptions = CompositeDisposable()
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/RoutingActivity.kt:23]: I'd set this in `onCreate` as well

File: app/src/main/java/net/squanchy/navigation/RoutingInjector.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import android.support.v7.app.AppCompatActivity
+
+import net.squanchy.injection.ActivityContextModule
+import net.squanchy.injection.ApplicationInjector
+import net.squanchy.navigation.deeplink.DeepLinkModule
+import net.squanchy.signin.SignInModule
+
+internal object RoutingInjector {
+
+    fun obtain(activity: AppCompatActivity): RoutingComponent {
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/RoutingInjector.kt]: this one doesn't need the object

File: app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkParsingException.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package net.squanchy.navigation.deeplink
+
+internal class DeepLinkParsingException private constructor(message: String) : RuntimeException(message) {
+
+    companion object {
+
+        fun eventDetailsMissingEventId(path: String): DeepLinkParsingException {
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkParsingException.kt:7]: these can be put as top level functions as well

File: app/src/main/java/net/squanchy/navigation/deeplink/RoutingException.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package net.squanchy.navigation.deeplink
+
+internal class RoutingException : RuntimeException {
+
+    private constructor(message: String) : super(message)
+
+    private constructor(message: String, cause: Throwable) : super(message, cause)
+
+    companion object {
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/deeplink/RoutingException.kt:9]: same here (putting reminders so they're easy to find)

File: app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        doIfAnimatedVectorDrawable(firstStartProgress.drawable) { it.start() }
+    }
+
+    override fun onStop() {
+        super.onStop()
+        connectivityManager.unregisterNetworkCallback(networkCallback)
+        doIfAnimatedVectorDrawable(firstStartProgress.drawable) { it.stop() }
+    }
+
+    private fun doIfAnimatedVectorDrawable(drawable: Drawable, action: (AnimatedVectorDrawable) -> Unit) {
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt]: usually these read nicer if you have `action: AnimatedVectorDrawable.() -> Unit`, then you can call it as `doIfAnimatedVectorDrawable(drawable) { stop() }`

not a blocker

File: app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    private fun continueToScheduleAfterDelay() {
+        firstStartProgress.postDelayed({
+            startActivity(continuationIntent)      // We don't use the navigator here, we basically want to restart the whole flow
+            finish()
+        }, DELAY_AFTER_ANIMATIONS_MILLIS)
+    }
+
+    private inner class NetworkConnectedCallback : ConnectivityManager.NetworkCallback() {
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt:149]: what's the difference between inner class and private top level classes?

**Review by fourlastor — CHANGES_REQUESTED** [review]
Cool stuff 😄 I'm not sure we need all the `internal` modifiers though, since this isn't even a lib :)

File: app/src/main/java/net/squanchy/navigation/Navigator.kt (hunk 4 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/Navigator.kt:35]: I would not hide the bitwise operation, that seems to me more confusing than anything else. It doesn't read any weirder than in Java...

File: app/src/main/java/net/squanchy/navigation/NavigationModule.kt (hunk 3 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/NavigationModule.kt:8]: Autoconverter did this, the compiler is fine, I am fine XD

File: app/src/main/java/net/squanchy/navigation/RoutingActivity.kt (hunk 5 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/RoutingActivity.kt:23]: I've moved it from there so this can be the `val` it deserves to be ✨

File: app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkParsingException.kt (hunk 7 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/deeplink/DeepLinkParsingException.kt:7]: I'd not be able to invoke the private ctor and I really don't want to escalate those ctors visibility

File: app/src/main/java/net/squanchy/navigation/deeplink/RoutingException.kt (hunk 8 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/deeplink/RoutingException.kt:9]: Ditto as above

File: app/src/main/java/net/squanchy/navigation/RoutingActivity.kt (hunk 5 above)
Comment (fourlastor) [reply app/src/main/java/net/squanchy/navigation/RoutingActivity.kt:23]: yes I know what you mean, as long as you're sure this doesn't mess with the lifecycle of the activity XD

File: app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt (hunk 10 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt:149]: https://kotlinlang.org/docs/reference/nested-classes.html

tl;dr it's a non-static inner class, because it can access the outer classes' fields (it's a direct translation from Java)

File: app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt (hunk 9 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt]: fd60dc074296f41dd74b5c8d6cfe3b8ccf3f7616

**Review by fourlastor — APPROVED** [review]

File: app/src/main/java/net/squanchy/navigation/RoutingActivity.kt (hunk 5 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/navigation/RoutingActivity.kt:23]: It's essentially the same thing it was before, it's just we don't usually leave initialisers on fields in Java but it's totally legit to do it in Kotlin (the subscription was effectively immutable in Java too)
